### PR TITLE
Only append default path to alertmanager.url config option if no path supplied

### PR DESCRIPTION
**Pull request by satterly**
Hardcoding the path of alertmanager to "/api/alerts" means it is difficult to
use the notification webhook for integrating tools other than alertmanager.

Now, if a path is supplied then it will be used in preference to the default.

### CODE REVIEW

**Comment by satterly** [comment]
I've just noticed that pull request #1003 conflicts with this change. If #1003 is going to be merged I can rebase my change against it. Let me know if I should go ahead or not. Thanks.

**Comment by fabxc** [comment]
No worries about that one. Closed it to avoid further confusion.

**Comment by satterly** [comment]
@fabxc I've rebased this against latest master. Any chance someone could have a look and give me some feedback? Thanks.

File: notification/notification.go [hunk 1]
```diff
@@ -31,7 +31,7 @@ import (
 )
 
 const (
-	alertPushEndpoint = "/api/v1/alerts"
+	AlertPushEndpoint = "/api/v1/alerts"
```
Comment (fabxc) [inline notification/notification.go]: This constant doesn't really have to be in that package any longer.
Let's just move it to an unexported constant in `cmd/prometheus/config.go`.

File: notification/notification_test.go [hunk 2]
```diff
@@ -89,8 +89,8 @@ func TestHandlerSend(t *testing.T) {
 	)
 
 	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		if r.URL.Path != alertPushEndpoint {
-			t.Fatalf("Bad endpoint %q used, expected %q", r.URL.Path, alertPushEndpoint)
+		if r.URL.Path != AlertPushEndpoint {
```
Comment (fabxc) [inline notification/notification_test.go]: As the full URL is now provided externally, this is no longer a behavior of the notification package.
We can remove that check.

File: notification/notification_test.go [hunk 3]
```diff
@@ -110,7 +110,7 @@ func TestHandlerSend(t *testing.T) {
 	defer server.Close()
 
 	h := New(&HandlerOptions{
-		AlertmanagerURL: server.URL,
+		AlertmanagerURL: server.URL + AlertPushEndpoint,
```
Comment (fabxc) [inline notification/notification_test.go]: And we can also remove this.

**Comment by fabxc** [comment]
Thanks, LGTM after fixes :+1:

**Comment by satterly** [comment]
@fabxc done!

File: cmd/prometheus/config.go [hunk 4]
```diff
@@ -198,7 +203,7 @@ func init() {
 
 	// Alertmanager.
 	cfg.fs.StringVar(
-		&cfg.notification.AlertmanagerURL, "alertmanager.url", "",
+		&cfg.alertmgrURL, "alertmanager.url", "",
 		"The URL of the alert manager to send notifications to.",
```
Comment (fabxc) [inline cmd/prometheus/config.go]: Can we add "(If no path is provided, the path is set to '/api/v1/alerts')"?

File: cmd/prometheus/config.go [hunk 5]
```diff
@@ -288,6 +297,25 @@ func parseInfluxdbURL() error {
...
+	if cfg.alertmgrURL == "" {
+		return nil
+	}
+
+	url, err := url.Parse(cfg.alertmgrURL)
+	if err != nil {
+		return err
+	}
+
+	if url.Path == "" {
```
Comment (brian-brazil) [inline cmd/prometheus/config.go]: Does this mean it's not possible to send alerts to http://example.com/  ?

This magic could be confusing for users.

Comment (satterly) [inline cmd/prometheus/config.go]: No, it's possible to send to any URL. For example...
```
$ ./prometheus -config.file=prometheus.yml -alertmanager.url=http://localhost:8080/
```
```
$ nc -l 8080
POST / HTTP/1.1
Host: localhost:8080
User-Agent: Go-http-client/1.1
Content-Length: 395
Content-Type: application/json
Accept-Encoding: gzip

[{"labels":{"alertname":"FooAlert","instance":"localhost:9090","job":"prometheus","tags":"one,two,three"},"annotations":{"description":"foo bar baz","severity":"critical"},"startsAt":"2015-12-30T11:46:52.804Z","endsAt":"0001-01-01T00:00:00Z","generatorURL":"http://macbookpro.home:9090/graph#%5B%7B%22expr%22%3A%22prometheus_local_storage_memory_chunkdescs%20%3E%200%22%2C%22tab%22%3A0%7D%5D"}]
```

**Comment by brian-brazil** [comment]
This integration is intended to only be used with the alertmanager, and we make no guarantees on the stability of this interface. If you want to hook into alerts the alertmanager webhook is provided for this purpose.

I don't think this is a feature we should be offering. If someone does want to try and pretend to be an alertmanager, it shouldn't be difficult for them to also handle the path of the endpoint we use.

I think that overall this change will create confusion for normal users and expose a hook in our system where we haven't planned to have one; for the sake of saving a small amount of code for a very advanced use case.

**Comment by satterly** [comment]
@brian-brazil ok, I can understand why you might not want to support an API that was only meant for internal prometheus use, but allowing someone to integrate with prometheus without insisting that they deploy alertmanager in addition to prometheus just to get a web hook working seemed like a win to me.

i can get what i want from a simple apache reverse proxy config...
```
Listen 9093
<VirtualHost *:9093>
     ProxyPass /api/v1/alerts http://localhost:8080/webhooks/prometheus
</VirtualHost>
```

**Comment by fabxc** [comment]
As discussed on other channels, I can still see good reasons to go either way. Leaving things be for now is certainly the safer option given that the whole integration is undergoing changes at the moment.

Restricting integration to Alertmanager is nothing I generally consider ideal. It may allow for a better user experience though.
It'll be worthwhile considering whether there is a clean solution satisfying either side once things have stabilized.

**Pull request closed without merging**